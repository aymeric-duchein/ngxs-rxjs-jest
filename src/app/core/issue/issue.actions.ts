import {Issue} from './issue.model';

export class OpenIssueModal {
  static readonly type = '[ISSUE] open issue modal';

  constructor(public issue: Issue) {
  }
}

export class HttpIssueSuccess {
  static readonly type = '[ISSUE] Http issue success';

  constructor(public issue: Issue) {
  }
}
