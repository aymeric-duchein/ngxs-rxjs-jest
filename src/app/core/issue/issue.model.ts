export enum IssueStatus {
  TODO = 'To do',
  IN_PROGRESS = 'In progress',
  DONE = 'Done',
}

export enum SortKey {
  NAME,
  DATE,
  STATUS,
  NONE
}

export enum SortDirection {
  ASC,
  DESC,
}

export enum IssueColor {
  color1 = '#323E40',
  color2 = '#F2AB27',
  color3 = '#D97D0D',
  color4 = '#732231',
  color5 = '#79BAF2',
  color6 = '#51608C',
  color7 = '#8697A6',
  color8 = '#BFCDD9',
  color9 = '#D9556D',
  color10 = '#4960A6',
  color11 = '#485922',
  color12 = '#798C35',
  color13 = '#B4BF5E',
  color14 = '#6B7FF2',
  color15 = '#F2E96D',
  color16 = '#F25C05',
  color17 = '#BF5934',
  color18 = '#BF1304',
}

export interface Issue {
  id: string;
  name: string;
  description: string;
  author: string;
  authorAvatar: string;
  date: string;
  status: IssueStatus;
  color: IssueColor;
}

export interface IssueSortCriteria {
  by: SortKey;
  direction: SortDirection;
}

export class IssueStateModel {
  issues: Issue[];
  sort?: IssueSortCriteria;
  search?: string;
}
