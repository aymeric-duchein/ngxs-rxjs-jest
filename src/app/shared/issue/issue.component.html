<mat-card (click)="emitIssue()">
  <mat-card-header>
    <div class="overlay">
      <mat-icon>edit</mat-icon>
    </div>
    <img mat-card-avatar [alt]="issue.authorAvatar" [src]="issue.authorAvatar">
    <mat-card-title>
      <span class="label" [style.background-color]="issue.color"></span>
      <span>{{ issue.name }}</span>
    </mat-card-title>
    <mat-card-subtitle>
      <span>{{issue.author}}</span>
      <span class="date"> - {{issue.date | date: 'dd/MM/yy'}}</span>
    </mat-card-subtitle>
    <div class="status-container">
      <div class="status">
        <span>{{ issue.status }}</span>
      </div>
    </div>
  </mat-card-header>
  <mat-card-content>
    <div class="description">{{ issue.description }}</div>
  </mat-card-content>
</mat-card>
