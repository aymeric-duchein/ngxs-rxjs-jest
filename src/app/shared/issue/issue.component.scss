@import "src/styles";

mat-card {
  margin: 0.7rem 0.4rem;
  position: relative;

  .overlay {
    position: absolute;
    background-color: rgba(230, 227, 227, 0.9);
    top: 16px;
    left: 16px;
    height: 40px;
    width: 40px;
    border-radius: 50%;
    opacity: 0;

    mat-icon {
      position: absolute;
      top: 9px;
      left: 10px;
      font-size: 22px;
      width: 22px;
      height: 22px;
      color: #a5a2a2;
      opacity: 0;
    }
  }

  mat-card-title {
    display: flex;
    align-items: center;
    font-size: 13px;

    .label {
      background-color: #b3bac5;
      border-radius: 4px;
      width: 2rem;
      margin-right: 0.5rem;
      display: block;
      height: 8px;
    }
  }

  mat-card-subtitle {
    font-size: 10px;

    .date {
      color: #c9c0c0;
    }
  }

  .status-container {
    position: absolute;
    top: 17px;
    right: 1rem;

    .status {
      font-size: 10px;
      background-color: rgba(#0A6ABF, 0.2);
      color: #0A6ABF;
      display: inline-flex;
      padding: 7px 12px;
      border-radius: 16px;
      align-items: center;
      height: 10px;
    }
  }

  mat-card-content {
    .description {
      font-size: 11px;
      height: 1rem;
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  }
}

mat-card:hover {
  background-color: #f4f5f7;

  .overlay {
    opacity: 1;
    transition: opacity .4s ease-in-out;

    mat-icon {
      opacity: 1;
      transition: opacity .3s ease-in-out;
    }
  }
}



