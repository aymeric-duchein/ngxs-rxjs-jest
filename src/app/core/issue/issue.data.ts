import {getRandomImg} from './issue.utils';
import {Issue, IssueColor, IssueStatus} from './issue.model';

export const fakeJsonUrl = 'https://app.fakejson.com/q';

export const DEFAULT_ISSUES: Issue[] = [
  {
    id: '1',
    name: 'name 1',
    description: 'zty',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.TODO,
    color: IssueColor.color16,
  },
  {
    id: '2',
    name: 'name 2',
    description: 'ztzeazey',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.DONE,
    color: IssueColor.color2,
  },
  {
    id: '3',
    name: 'name 1',
    description: 'zty',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.TODO,
    color: IssueColor.color14,
  },
  {
    id: '4',
    name: 'name 2',
    description: 'ztzeazey',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.DONE,
    color: IssueColor.color15,
  },
  {
    id: '5',
    name: 'name 1',
    description: 'zty',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.IN_PROGRESS,
    color: IssueColor.color5,
  },
  {
    id: '6',
    name: 'name 2',
    description: 'ztzeazey',
    author: 'Aymeric',
    authorAvatar: getRandomImg(),
    date: new Date().toISOString(),
    status: IssueStatus.IN_PROGRESS,
    color: IssueColor.color12,
  },
];
