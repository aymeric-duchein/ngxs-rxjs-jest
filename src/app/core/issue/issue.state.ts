import {Action, createSelector, Selector, State, StateContext} from '@ngxs/store';
import {Issue, IssueSortCriteria, IssueStateModel, IssueStatus, SortDirection} from './issue.model';
import {BoardIssueClick, BoardIssueDropped} from '../../board/board.actions';
import {ListingIssueClick, SearchFormUpdated, SortOptionClick} from '../../listing/listing.actions';
import {SnackbarUndoIssueUpdate} from '../snackbar/snackbar.actions';
import {AddHttpIssueBtnClick, AddIssueBtnClick} from '../header/header.action';
import {IssueUpsertClosed} from '../modal/modal.action';
import {HttpIssueSuccess, OpenIssueModal} from './issue.actions';
import {by, filterBy, filterByStatus, getFakeJsonPayload, getRandomColor} from './issue.utils';
import {DEFAULT_ISSUES, fakeJsonUrl} from './issue.data';
import {HttpClient} from '@angular/common/http';
import {map, switchMap} from 'rxjs/operators';

export const ISSUE_STATE_NAME = 'issue';

@State<IssueStateModel>({
  name: ISSUE_STATE_NAME,
  defaults: {
    issues: DEFAULT_ISSUES
  }
})
export class IssueState {

  constructor(private readonly httpClient: HttpClient) {
  }

  @Selector()
  static getIssues(state: IssueStateModel): Issue[] {
    return [...state.issues];
  }

  @Selector()
  static getSortCriteria(state: IssueStateModel): IssueSortCriteria {
    return state.sort;
  }

  @Selector()
  static getFilter(state: IssueStateModel): string {
    return state.search;
  }

  @Selector([IssueState.getIssues, IssueState.getSortCriteria])
  static getSortedIssues(_, issues: Issue[], criteria: IssueSortCriteria): Issue[] {
    return issues.sort(by(criteria));
  }

  @Selector([IssueState.getIssues, IssueState.getFilter])
  static getFilteredIssues(_, issues: Issue[],  search: string): Issue[] {
    return issues.filter(filterBy(search));
  }

  @Selector([IssueState.getSortedIssues, IssueState.getFilter])
  static getFilteredAndSortedIssues(_, sortedIssues: Issue[], search: string): Issue[] {
    return sortedIssues.filter(filterBy(search));
  }

  static getIssuesByStatus(status: IssueStatus) {
    return createSelector([IssueState], ({issues}: IssueStateModel) => issues.filter(filterByStatus(status)));
  }

  @Action([BoardIssueClick, ListingIssueClick, AddIssueBtnClick])
  openIssueModal({dispatch}, action: AddIssueBtnClick | ListingIssueClick | BoardIssueClick) {
    return dispatch(new OpenIssueModal((action as any).issue));
  }

  @Action(AddHttpIssueBtnClick)
  getHttpIssue({dispatch}: StateContext<IssueStateModel>) {
    const payload = getFakeJsonPayload();
    return this.httpClient.post(fakeJsonUrl, payload).pipe(
      map((response: Issue) => {
        const issue: Issue = {
          id: response.id,
          name: response.name,
          description: response.description,
          author: response.author,
          authorAvatar: response.authorAvatar,
          date: response.date,
          status: response.status,
          color: getRandomColor(),
        };
        return issue;
      }),
      switchMap(issue => dispatch(new HttpIssueSuccess(issue)))
    );
  }

  @Action(BoardIssueDropped)
  issuesDropped({patchState, getState}: StateContext<IssueStateModel>, {droppedIssue, status}: BoardIssueDropped) {
    const issues = getState().issues;
    patchState({issues: [...issues.filter(i => i.id !== droppedIssue.id), {...droppedIssue, status}]});
  }

  @Action(SearchFormUpdated)
  searchUpdated({patchState}: StateContext<IssueStateModel>, {search}: SearchFormUpdated) {
    patchState({search});
  }

  @Action(SortOptionClick)
  setSortOption({patchState, getState}: StateContext<IssueStateModel>, {sortKey}: SortOptionClick) {
    const {sort} = getState();
    if (sort && sort.by === sortKey) {
      patchState({sort: {by: sortKey, direction: sort.direction === SortDirection.ASC ? SortDirection.DESC : SortDirection.ASC }});
    } else {
      patchState({sort: {by: sortKey, direction: SortDirection.ASC}});
    }
  }

  @Action(IssueUpsertClosed)
  upsertIssue({patchState, getState}: StateContext<IssueStateModel>, {issue}: IssueUpsertClosed) {
    const {issues} = getState();
    patchState({issues: [...issues.filter(i => i.id !== issue.id), issue]});
  }

  @Action(HttpIssueSuccess)
  saveIssue({patchState, getState}: StateContext<IssueStateModel>, {issue}: HttpIssueSuccess) {
    const {issues} = getState();
    patchState({issues: [...issues.filter(i => i.id !== issue.id), issue]});
  }

  @Action(SnackbarUndoIssueUpdate)
  resetIssueToPreviousState({patchState, getState}: StateContext<IssueStateModel>, {issue}: SnackbarUndoIssueUpdate) {
    const {issues} = getState();
    patchState({issues: [...issues.filter(i => i.id !== issue.id), issue]});
  }
}
