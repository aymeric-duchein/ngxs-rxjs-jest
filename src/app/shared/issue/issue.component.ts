import {Component, EventEmitter, Input, Output} from '@angular/core';
import {Issue} from '../../core/issue/issue.model';

@Component({
  selector: 'app-issue',
  templateUrl: './issue.component.html',
  styleUrls: ['./issue.component.scss']
})
export class IssueComponent {

  @Input() issue: Issue;

  @Output() issueClick: EventEmitter<Issue> = new EventEmitter<Issue>();

  emitIssue() {
    this.issueClick.emit(this.issue);
  }
}
