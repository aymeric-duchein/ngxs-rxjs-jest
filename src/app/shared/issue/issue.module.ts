import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {IssueComponent} from './issue.component';
import {MatCardModule, MatIconModule} from '@angular/material';
import {HttpClientModule} from '@angular/common/http';

@NgModule({
  declarations: [IssueComponent],
  exports: [IssueComponent],
  imports: [
    CommonModule,
    MatCardModule,
    MatIconModule,
    HttpClientModule,
  ],
})
export class IssueModule {
}
