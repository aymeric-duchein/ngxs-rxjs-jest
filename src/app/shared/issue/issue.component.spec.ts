import {ComponentFixture, TestBed} from '@angular/core/testing';

import {IssueComponent} from './issue.component';
import {MatCardModule, MatIconModule} from '@angular/material';
import {MockModule, MockRender} from 'ng-mocks';
import {MOCK_TODO_ISSUE} from '@mocks/issue.mock';
import {By} from '@angular/platform-browser';
import {Issue} from '@src/app/core/issue/issue.model';

describe('IssueComponent', () => {
  let component: { issue: Issue, issueClick: jest.Mock };
  let fixture: ComponentFixture<{ issue: Issue, issueClick: jest.Mock }>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [IssueComponent],
      imports: [
        MockModule(MatCardModule),
        MockModule(MatIconModule)
      ]
    });

    fixture = MockRender('<app-issue [issue]="issue" (issueClick)="issueClick($event)"></app-issue>',
      {
        issue: MOCK_TODO_ISSUE,
        issueClick: jest.fn()
      }
    );
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should emit event on click', () => {
    const card = fixture.debugElement.query(By.css('mat-card'));
    card.triggerEventHandler('click', {});
    fixture.detectChanges();

    expect(component.issueClick).toHaveBeenCalledWith(MOCK_TODO_ISSUE);
  });
});
