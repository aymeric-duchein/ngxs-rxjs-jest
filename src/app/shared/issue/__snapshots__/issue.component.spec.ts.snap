// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueComponent should match snapshot 1`] = `
<mock-render
  issue={[Function Object]}
  issueClick={[Function Function]}
>
  <app-issue
    ng-reflect-issue="[object Object]"
  >
    <mat-card>
      <mat-card-header>
        <div
          class="overlay"
        >
          <mat-icon>
            edit
          </mat-icon>
        </div>
        <img
          alt="fakeUrl"
          mat-card-avatar=""
          src="fakeUrl"
        />
        <mat-card-title>
          <span
            class="label"
            style="background-color: rgb(242, 92, 5);"
          />
          <span>
            name 1
          </span>
        </mat-card-title>
        <mat-card-subtitle>
          <span>
            Author fake
          </span>
          <span
            class="date"
          >
             - 05/01/20
          </span>
        </mat-card-subtitle>
        <div
          class="status-container"
        >
          <div
            class="status"
          >
            <span>
              To do
            </span>
          </div>
        </div>
      </mat-card-header>
      <mat-card-content>
        <div
          class="description"
        >
          Fake description
        </div>
      </mat-card-content>
    </mat-card>
  </app-issue>
</mock-render>
`;
